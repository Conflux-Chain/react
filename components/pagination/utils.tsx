export const getPageCount = (total: number, pageSize: number) => {
  return Math.floor((total - 1) / pageSize) + 1
}
