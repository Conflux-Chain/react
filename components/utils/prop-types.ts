export const tuple = <T extends string[]>(...args: T) => args
//todo better syntax related to tuple?
export const variantsTuple = <T extends ('solid' | 'line' | 'text')[]>(...args: T) => args
export const statusTuple = <T extends ('default' | 'hover' | 'active' | 'disabled')[]>(
  ...args: T
) => args

const buttonVariants = variantsTuple('solid', 'line', 'text')
const tabVariants = variantsTuple('solid', 'line')
const inputVariants = tuple('solid', 'line')

const buttonColors = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const tagColors = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const buttonStatus = tuple('default', 'hover', 'active', 'disabled')
const tabStatus = statusTuple('default', 'hover', 'active', 'disabled')

const messageColors = tuple('default', 'primary', 'success', 'warning', 'error')

const selectVariants = tuple('line', 'text')

const normalSizes = tuple('mini', 'small', 'medium', 'large')

const tagSizes = tuple('small', 'medium', 'large')

const normalTypes = tuple('default', 'primary', 'secondary', 'success', 'warning', 'error')

const inputColors = tuple('default', 'primary', 'success', 'warning', 'error')

const themeTypes = tuple('dark', 'light')

const snippetColors = tuple('default', 'secondary', 'success', 'warning', 'error', 'dark', 'lite')

const cardVariants = tuple('solid', 'line')

const cardColors = tuple('default', 'primary', 'success', 'warning', 'error')

const copyTypes = tuple('default', 'slient', 'prevent')

const triggerTypes = tuple('hover', 'click')

const placement = tuple(
  'top',
  'topStart',
  'topEnd',
  'left',
  'leftStart',
  'leftEnd',
  'bottom',
  'bottomStart',
  'bottomEnd',
  'right',
  'rightStart',
  'rightEnd',
)

const dividerAlign = tuple('start', 'center', 'end', 'left', 'right')

const paginationVariants = tuple('line', 'solid')

export type ButtonColors = typeof buttonColors[number]

export type TagColors = typeof tagColors[number]

export type ButtonVariants = typeof buttonVariants[number]
export type TabVarient = typeof tabVariants[number]

export type TagVariants = ButtonVariants

export type ButtonStatus = typeof buttonStatus[number]
export type TabStatus = typeof tabStatus[number]

export type MessageColors = typeof messageColors[number]

export type NormalSizes = typeof normalSizes[number]

export type TagSizes = typeof tagSizes[number]

export type SelectVariants = typeof selectVariants[number]

export type NormalTypes = typeof normalTypes[number]

export type InputColors = typeof inputColors[number]

export type InputVariantTypes = typeof inputVariants[number]

export type ThemeTypes = typeof themeTypes[number]

export type SnippetColors = typeof snippetColors[number]

export type CardColors = typeof cardColors[number]

export type CardVariants = typeof cardVariants[number]

export type CopyTypes = typeof copyTypes[number]

export type TriggerTypes = typeof triggerTypes[number]

export type Placement = typeof placement[number]

export type DividerAlign = typeof dividerAlign[number]

export type PaginationVariants = typeof paginationVariants[number]
