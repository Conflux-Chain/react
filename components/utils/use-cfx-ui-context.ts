import React from 'react'
import { ToastWithID } from '../use-toasts/toast-container'
import { MessageItemProps } from '../use-messages/message-item'
import { NotificationItemProps } from '../use-notifications/notification-item'

export type UpdateToastsFunction<T> = (fn: (toasts: Array<T>) => Array<T>) => any
export type UpdateMessagesFunction<T> = (fn: (messages: Array<T>) => Array<T>) => any
export type UpdateNotificationsFunction<T> = (fn: (notifications: Array<T>) => Array<T>) => any

export interface CfxUIContextParams {
  messages: Array<MessageItemProps>
  updateMessages: UpdateMessagesFunction<MessageItemProps>
  notifications: Array<NotificationItemProps>
  updateNotifications: UpdateNotificationsFunction<NotificationItemProps>
  toasts: Array<ToastWithID>
  toastHovering: boolean
  updateToasts: UpdateToastsFunction<ToastWithID>
  updateToastHoverStatus: Function
}

const defaultParams: CfxUIContextParams = {
  messages: [],
  updateMessages: t => t,
  notifications: [],
  updateNotifications: t => t,
  toasts: [],
  toastHovering: false,
  updateToasts: t => t,
  updateToastHoverStatus: () => {},
}

export const CfxUIContent: React.Context<CfxUIContextParams> = React.createContext<
  CfxUIContextParams
>(defaultParams)

export const useCfxUIContext = (): CfxUIContextParams =>
  React.useContext<CfxUIContextParams>(CfxUIContent)
