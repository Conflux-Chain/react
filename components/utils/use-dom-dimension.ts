import React, { useEffect, useRef, useState } from 'react'

export default function useDOMDimension<E extends HTMLElement>(
  property: 'clientWidth' | 'offsetWidth' = 'clientWidth', //more to come
  ...gurads: any[]
): [number, React.RefObject<E>] {
  const ref = useRef<E>(null)
  const [width, setWidth] = useState(0)
  useEffect(() => {
    if (ref.current) {
      setWidth(ref.current[property])
    }
  }, [...gurads])
  return [width, ref]
}
