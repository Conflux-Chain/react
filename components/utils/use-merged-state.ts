import { useRef, useState, useCallback } from 'react'
// https://github.com/react-component/util/blob/4045d568885d65a5491ab22347682f5e31da31dc/src/hooks/useMergedState.ts
export default function useControlledState<T>(
  defaultValue: T,
  option?: {
    value?: T
    onChange?: (value: T, prevValue: T) => void
    postState?: (value: T) => T
  },
): [T, (value: T) => void] {
  const { value, onChange, postState } = option || {}
  const [innerValue, setInnerValue] = useState<T>(value !== undefined ? value : defaultValue)

  // controlled take priority of uncontrolled
  let calcualteValue = value !== undefined ? value : innerValue
  if (postState) {
    calcualteValue = postState(calcualteValue)
  }

  // value inside setValue should be updated
  const mergedValue = useRef(calcualteValue)
  mergedValue.current = calcualteValue

  // do not change the identity of setValue
  const setValue = useCallback((newValue: T) => {
    if (!value) {
      console.log('setValue', newValue)
      setInnerValue(newValue)
    }
    if (mergedValue.current !== newValue && onChange) {
      onChange(newValue, mergedValue.current)
    }
  }, [])

  return [mergedValue.current, setValue]
}
