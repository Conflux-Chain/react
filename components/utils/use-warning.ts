const warningStack: { [key: string]: boolean } = {}

const useWarning = (
  message: string,
  component?: string,
  level?: 'error' | 'warn' | 'info' | 'log',
) => {
  const tag = component ? ` [${component}]` : ' '
  const log = `[@cfxjs/react-ui]${tag}: ${message}`

  if (typeof console === 'undefined') return
  if (warningStack[log]) return
  warningStack[log] = true

  if (level) return console[level](log)

  if (process.env.NODE_ENV !== 'production') {
    return console.error(log)
  }

  console.warn(log)
}

export default useWarning
