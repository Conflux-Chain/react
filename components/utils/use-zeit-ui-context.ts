import React from 'react'
import { ToastWithID } from '../use-toasts/toast-container'
import { MessageWithID } from '../use-messages/toast-container'

export type UpdateToastsFunction<T> = (fn: (toasts: Array<T>) => Array<T>) => any
export type UpdateMessagesFunction<T> = (fn: (messages: Array<T>) => Array<T>) => any

export interface ZeitUiContextParams {
  messages: Array<MessageWithID>
  updateMessages: UpdateMessagesFunction<ToastWithID>
  toasts: Array<ToastWithID>
  toastHovering: boolean
  updateToasts: UpdateToastsFunction<ToastWithID>
  updateToastHoverStatus: Function
}

const defaultParams: ZeitUiContextParams = {
  messages: [],
  updateMessages: t => t,
  toasts: [],
  toastHovering: false,
  updateToasts: t => t,
  updateToastHoverStatus: () => {},
}

export const ZEITUIContent: React.Context<ZeitUiContextParams> = React.createContext<
  ZeitUiContextParams
>(defaultParams)

export const useZEITUIContext = (): ZeitUiContextParams =>
  React.useContext<ZeitUiContextParams>(ZEITUIContent)
