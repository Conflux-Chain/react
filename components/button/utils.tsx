import React, { ReactNode } from 'react'
import { NormalSizes } from '../utils/prop-types'
import ButtonIcon from './button-icon'
import { ButtonProps } from './button'
import { ButtonGroupConfig } from '../button-group/button-group-context'

export const getButtonChildrenWithIcon = (
  loading: boolean,
  auto: boolean,
  size: NormalSizes,
  children: ReactNode,
  icons: {
    icon?: React.ReactNode
    iconRight?: React.ReactNode
  },
  isTag: boolean,
) => {
  const { icon, iconRight } = icons
  const hasIcon = icon || iconRight
  const isRight = Boolean(iconRight)
  const paddingForAutoMode =
    auto || size === 'mini' || isTag
      ? `calc(var(--zeit-ui-button-height) / 2 + var(--zeit-ui-button-padding) * .5)`
      : 0
  if (!hasIcon) return <div className={`text ${loading ? 'hidden' : ''}`}>{children}</div>
  return (
    <>
      <ButtonIcon isRight={isRight}>{hasIcon}</ButtonIcon>
      <div className={`text ${isRight ? 'right' : 'left'} ${loading ? 'hidden' : ''}`}>
        {children}
        <style jsx>{`
          .left {
            padding-left: ${paddingForAutoMode};
          }
          .right {
            padding-right: ${paddingForAutoMode};
          }
        `}</style>
      </div>
    </>
  )
}

export const filterPropsWithGroup = (
  props: React.PropsWithChildren<ButtonProps>,
  config: ButtonGroupConfig,
): ButtonProps => {
  if (!config.isButtonGroup) return props
  return {
    ...props,
    auto: true,
    shadow: false,
    ghost: config.ghost || props.ghost,
    size: config.size || props.size,
    variant: config.variant || props.variant,
    color: config.color || props.color,
    disabled: config.disabled || props.disabled,
  }
}
