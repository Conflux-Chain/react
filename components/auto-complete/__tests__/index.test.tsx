import AutoComplete, { useAutoCompleteHandle } from '../'
import { mount, render } from 'enzyme'
import React, { useEffect } from 'react'
import { act } from 'react-dom/test-utils'
import { nativeEvent, updateWrapper } from 'tests/utils'

const simulateNativeClick = (el: Element) => {
  el.dispatchEvent(
    new MouseEvent('click', {
      view: window,
      bubbles: true,
      cancelable: true,
    }),
  )
}

const typeWrapper = mount(<AutoComplete />)
typeWrapper.unmount()
const clickAway = async (wrapper: typeof typeWrapper) => {
  await act(async () => {
    simulateNativeClick(document.body)
  })
  await updateWrapper(wrapper, 150)
}
const changeTo = async (wrapper: typeof typeWrapper, value: string) => {
  wrapper
    .find('input')
    .at(0)
    .simulate('change', { ...nativeEvent, target: { value } })
}

describe('AutoComplete', () => {
  describe('Style', () => {
    it('should render correctly', () => {
      const wrapper = mount(<AutoComplete />)
      expect(() => wrapper.unmount()).not.toThrow()
      expect(
        render(
          <div>
            <AutoComplete />
            <AutoComplete variant="solid" />
            <AutoComplete clearable />
            <AutoComplete disabled />
          </div>,
        ),
      ).toMatchSnapshot()
    })

    it('should render with right variants, sizes and color', () => {
      const wrapper = mount(
        <div>
          <AutoComplete variant="solid" />
          <AutoComplete color="primary" />
          <AutoComplete color="success" />
          <AutoComplete size="mini" />
          <AutoComplete size="large" />
        </div>,
      )
      expect(() => wrapper.unmount()).not.toThrow()
    })

    it('should response width change', async () => {
      const wrapper = mount(<AutoComplete defaultValue="value" width="100px" />)
      expect(wrapper.prop('width')).toEqual('100px')
      await act(async () => {
        wrapper.setProps({ width: '200px' })
      })

      expect(wrapper.prop('width')).toEqual('200px')
    })
  })

  describe('Value uncontrolled', () => {
    it('should set input value from initial value', () => {
      let wrapper = mount(<AutoComplete defaultValue="value" />)
      let input = wrapper.find('input').at(0).getDOMNode()
      expect((input as HTMLInputElement).value).toEqual('value')

      wrapper = mount(<AutoComplete value="value2" />)
      input = wrapper.find('input').at(0).getDOMNode()
      expect((input as HTMLInputElement).value).toEqual('value2')
    })

    it('should work with useAutoCompleteHandle', () => {
      let log = ''
      jest.spyOn(console, 'log').mockImplementation(msg => (log = msg))
      function MockAutoComplete({ testSetValue = false, testGetValue = false }) {
        const { ref, getValue, setValue } = useAutoCompleteHandle()
        useEffect(() => {
          testGetValue && console.log(getValue())
        }, [testGetValue])
        useEffect(() => {
          testSetValue && setValue('foo')
        }, [testSetValue])
        return <AutoComplete ref={ref} defaultValue="value" />
      }

      const wrapper = mount(<MockAutoComplete />)
      wrapper.setProps({ testGetValue: true })
      expect(log).toEqual('value')
      wrapper.setProps({ testSetValue: true })
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('foo')
    })

    it('should set input value by select', () => {
      const focusHandler = jest.fn()
      let wrapper = mount(<AutoComplete options={[{ label: 'foo' }]} onFocus={focusHandler} />)
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      expect(focusHandler).toBeCalledTimes(1)
      wrapper.find('.auto-complete-dropdown .item').at(0).simulate('click', nativeEvent)
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('foo')
    })

    it('should work with prop clearable', async () => {
      const onClear = jest.fn()
      const wrapper = mount(<AutoComplete defaultValue="value" />)
      expect(wrapper.find('svg').length).toBe(0)

      await act(async () => {
        wrapper.setProps({ clearable: true, onClearClick: onClear })
      })
      expect(wrapper.find('svg').length).toBe(1)

      wrapper.find('svg').simulate('click', nativeEvent)
      expect(onClear).toBeCalledTimes(1)
      const input = wrapper.find('input').at(0).getDOMNode()
      expect((input as HTMLInputElement).value).toEqual('')
    })

    it('should reset the value when clicked away', async () => {
      const blurHandler = jest.fn()
      const wrapper = mount(
        <AutoComplete
          defaultValue="foo"
          disableFreeSolo
          onBlur={blurHandler}
          options={[{ label: 'foo' }, { label: 'bar' }]}
        />,
      )

      changeTo(wrapper, 'value')
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('value')

      await clickAway(wrapper)
      expect(blurHandler).toBeCalledTimes(1)

      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('foo')
    })

    it('should remain blank when clicked away', async () => {
      const wrapper = mount(
        <AutoComplete
          defaultValue="foo"
          disableFreeSolo
          options={[{ label: 'foo' }, { label: 'bar' }]}
        />,
      )

      changeTo(wrapper, '')
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('')

      await clickAway(wrapper)

      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('')
    })
  })

  describe('Value controlled', () => {
    it('should be render the controlled value', () => {
      const wrapper = mount(<AutoComplete value="value" />)
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('value')
      wrapper.setProps({ value: 'foo' })
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('foo')
    })

    it('should not set input value by select', () => {
      let wrapper = mount(<AutoComplete variant="solid" value="bar" options={[{ label: 'foo' }]} />)
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      wrapper.find('.auto-complete-dropdown .item').at(0).simulate('click', nativeEvent)
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('bar')
    })

    it('should not clear the input with prop clearable', async () => {
      const wrapper = mount(<AutoComplete value="value" clearable />)
      wrapper.find('svg').simulate('click', nativeEvent)
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('value')
    })

    it('should not reset input value when clicked away', async () => {
      const wrapper = mount(<AutoComplete value="value" />)

      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      changeTo(wrapper, 'foo')
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('value')

      await clickAway(wrapper)
      expect((wrapper.find('input').at(0).getDOMNode() as HTMLInputElement).value).toEqual('value')
    })
  })

  describe('Open uncontrolled', () => {
    it('should dropdown visible logic should be uncontrolled', async () => {
      // default close
      let wrapper = mount(<AutoComplete options={[{ label: 'foo' }]} />)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      // focus to open
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      // click away to close
      await clickAway(wrapper)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      // change to open
      changeTo(wrapper, 'bar')
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      // click away to close
      await clickAway(wrapper)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      // click to open
      wrapper.find('input').at(0).simulate('click', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      wrapper.unmount()
    })

    it('should close the dropdown when select', async () => {
      let wrapper = mount(<AutoComplete options={[{ label: 'foo' }]} />)
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      wrapper.find('.auto-complete-dropdown .item').at(0).simulate('click', nativeEvent)
      await updateWrapper(wrapper, 150)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
    })
  })

  describe('Open controlled', () => {
    it('should remain open', async () => {
      let wrapper = mount(<AutoComplete open options={[{ label: 'foo' }]} />)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      await clickAway(wrapper)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      wrapper.find('input').at(0).simulate('blur', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
      wrapper.setProps({ open: false })
      await updateWrapper(wrapper, 150)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      changeTo(wrapper, 'bar')
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      wrapper.find('input').at(0).simulate('click', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(0)
    })

    it('should not close the dropdown when select', () => {
      let wrapper = mount(<AutoComplete open options={[{ label: 'foo' }]} />)
      wrapper.find('input').at(0).simulate('focus', nativeEvent)
      wrapper.find('.auto-complete-dropdown .item').at(0).simulate('click', nativeEvent)
      expect(wrapper.find('.auto-complete-dropdown .item').length).toEqual(1)
    })
  })
})
