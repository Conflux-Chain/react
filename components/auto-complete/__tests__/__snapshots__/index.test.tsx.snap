// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoComplete should render correctly 1`] = `
<div>
  <AutoComplete
    className=""
    clearable={false}
    disableFreeSolo={false}
    disableMatchWidth={false}
    disabled={false}
    initialValue=""
    options={Array []}
    size="medium"
    solid={false}
  />
  <AutoComplete
    className=""
    clearable={false}
    disableFreeSolo={false}
    disableMatchWidth={false}
    disabled={false}
    initialValue=""
    options={Array []}
    size="medium"
    solid={true}
  />
</div>
`;
